File: Go/fls/parsefile/zip.go
package parsefile

import (
	"bytes"
	"encoding/binary"
	"io"
)

const (
	cNO        = 0
	cDeflate   = 8
	cDeflate64 = 9
	cBZIP2     = 12
	cLZMA      = 14
	cPPMd      = 98
)

const (
	eNO      = 0
	eUNKNOWN = 1
	eAES128  = 2
	eAES192  = 3
	eAES256  = 4
)

type Infzip struct {
	Hpos  uint64
	Hsize uint16

	Encryption byte
	Method     uint16
	CRC32      [4]byte
	Dsize      uint32
	Fsize      uint32
	Fname      string
}

func readHdr(rs io.ReadSeeker) *Infzip {
	hdr := make([]byte, 26)
	_, err := rs.Read(hdr)
	if err != nil {
		return nil
	}
	var res Infzip
	offset, _ := rs.Seek(0, io.SeekCurrent)
	res.Hpos = uint64(offset - 26)
	res.Hsize = 26

	res.Encryption = hdr[2] & 1
	res.Method = binary.LittleEndian.Uint16(hdr[4:6])
	copy(res.CRC32[:], hdr[10:14])
	res.Dsize = binary.LittleEndian.Uint32(hdr[14:18])
	res.Fsize = binary.LittleEndian.Uint32(hdr[18:22])

	szfn := binary.LittleEndian.Uint16(hdr[22:24])
	res.Hsize += szfn
	fns := make([]byte, szfn)
	_, err = rs.Read(fns)
	if err != nil {
		return nil
	}
	res.Fname = string(fns)

	szex := binary.LittleEndian.Uint16(hdr[24:26])
	if szex != 0 {
		res.Hsize += szex
		ext := make([]byte, szex)
		_, err = rs.Read(ext)
		if err != nil {
			return nil
		}
		if res.Method == 99 {
			res.Encryption = ext[8] + 1
			res.Method = binary.LittleEndian.Uint16(ext[9:11])
		}
	}
	return &res
}

func ReadInfzip(rs io.ReadSeeker) []Infzip {
	var res []Infzip
	hdr := make([]byte, 4)
	for {
		_, err := rs.Read(hdr)
		if err != nil {
			break
		}
		if bytes.Equal(hdr, []byte{0x50, 0x4b, 0x03, 0x04}) {
			inf := readHdr(rs)
			if inf == nil {
				break
			}
			//folder
			if inf.Fname[len(inf.Fname)-1] == '/' {
				continue
			}
			res = append(res, *inf)
			rs.Seek(int64(inf.Dsize), io.SeekCurrent)
		} else {
			break
		}
	}
	return res
}

File: Python/fls/parse/zip.py
comprType = {"NO":0, "Deflate":8, "Deflate64":9, "BZIP2":12, "LZMA":14, "PPMd":98}
encrType = {"NO":0, "UNKNOWN":1, "AES128":2, "AES192":3, "AES256":4}

def read_inf(fl):
	pos = 0
	while True:
		fl.seek(pos, 0)
		h = fl.read(4)
		if h == b"\x50\x4b\x03\x04":
			hdr = fl.read(26)
			if hdr == b'':
				break
			pos += 4

			res = {}
			res["hpos"] = pos
			res["hsize"] = 26
			res["encryption"] = hdr[2] & 1
			res["method"] = int.from_bytes(hdr[4:6], byteorder='little')
			res["crc32"] = hdr[10:14]
			res["dsize"] = int.from_bytes(hdr[14:18], byteorder='little')
			res["fsize"] = int.from_bytes(hdr[18:22], byteorder='little')

			szfn = int.from_bytes(hdr[22:24], byteorder='little')
			res["hsize"] += szfn
			res["fname"] = fl.read(szfn)

			szex = int.from_bytes(hdr[24:26], byteorder='little')
			if szex != 0:
				res["hsize"] += szex
				ext = fl.read(szex)
				if res["method"] == 99:
					res["encryption"] = ext[8] + 1
					res["method"] = int.from_bytes(ext[9:11], byteorder='little')
			pos += res["hsize"] + res["dsize"]
			yield res
		else:
			break
